### PR TITLE
NOTICK - Eliminate unneeded code

### PR DESCRIPTION
**Pull request by kyriathar**

### CODE REVIEW

**Comment by lankydan** [comment]
If we are just removing code, doesn't that mean the race condition is still there? Is there a way to fix it?

**Comment by kyriathar** [comment]
The race condition is resolved by moving the registering to `DBTransactionStorage.updates` prior to calling `getTransaction` in `DBTransactionStorage.trackTransaction`.

**Comment by blsemo** [comment]
Not sure I quite follow your comment - is the change to remove the race condition also in this PR?

**Comment by kyriathar** [comment]
This was done by @josephzunigadaly when implementing https://r3-cev.atlassian.net/browse/ENT-4857. I left that bit of code (in `DBTransactionStorage.trackTransaction` and the related tests) untouched yes.

File: node/src/main/kotlin/net/corda/node/services/persistence/DBTransactionStorage.kt [hunk 1]
```diff
@@ -223,31 +219,11 @@ class DBTransactionStorage(private val database: CordaPersistence, cacheFactory:
...
-        return Pair(trackTransactionWithNoWarning(id), warning)
-    }
-
-    override fun trackTransactionWithNoWarning(id: SecureHash): CordaFuture<SignedTransaction> {
         val updateFuture = updates.filter { it.id == id }.toFuture()
         return database.transaction {
             txStorage.locked {
                 val existingTransaction = getTransaction(id)
                 if (existingTransaction == null) {
-                    updates.filter { it.id == id }.toFuture()
```
Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/services/persistence/DBTransactionStorage.kt:250]: I think this line of the PR is valid (I went back to look at @josephzunigadaly original and this accidentally got left in the OS port).

However, I don't think we can do the rest of it.  The `trackTransaction` is public API which can be called by CorDapp developers inside a database transaction (which triggers the race condition, and hence the warning).

**Review by rick-r3 — CHANGES_REQUESTED** [review]

**Comment by rick-r3** [comment]
Actually I saw your slack comment about it not being public API.  Maybe that is true....

File: node/src/main/kotlin/net/corda/node/services/persistence/DBTransactionStorage.kt (hunk 1 above)
Comment (lankydan) [reply node/src/main/kotlin/net/corda/node/services/persistence/DBTransactionStorage.kt:250]: `trackTransaction` is public, the with warning is not.

**Comment by rick-r3** [comment]
... ah I was blind `val validatedTransactions: TransactionStorage` on `ServiceHub`.  It is public API.

**Comment by lankydan** [comment]
> ... ah I was blind `val validatedTransactions: TransactionStorage` on `ServiceHub`. It is public API.

Yes, but only an internal function was removed (since it was on `WritableTransactionStorage`, not `TransactionStorage`).

**Comment by rick-r3** [comment]
No my argument was more about the fact we can't remove the warning because you can trigger via public API.  Now, technically I think we can (and I hadn't realised) because we think we're moving the default isolation level to READ_COMMITTED which supposedly stops the race, but I think it is actually possible to still set a different isolation level via the Hikari data source properties.  So 🤷‍♂️

**Comment by josephzunigadaly** [comment]
I think we should undo the changes that remove the warning and keep the removal of `updates.filter { it.id == id }.toFuture()`.

**Comment by kyriathar** [comment]
Reverted -removing the warning- change, however would it make it better if we also check if the existing transaction when calling `trackTransaction` is `REPEATABLE_READ`? I think that's the only case we can potentially get a race due to transaction isolation level if they executed a `SELECT` before.

**Review by josephzunigadaly — DISMISSED** [review]

**Review by josephzunigadaly — APPROVED** [review]

**Pull request merged**